students/Dustin_L/lesson09/mailroom4.py

```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+"""Mail Room 4 Module
+
+This module contains all of the functions for the updated Mail Room 4 module.
+"""
+
+import datetime
+from collections import defaultdict
+
+SELECT_PROMPT = ('\nPlease select from the following options:\n'
+                 '\t1. Send a Thank You\n'
+                 '\t2. Create a Report\n'
+                 '\t3. Send letters to all donors\n'
+                 '\t4. quit\n'
+                 ' --> ')
+PROMPT_OPTS = (1, 2, 3, 4)
+
+
+class Donor:
+    """Contains all information for a single donor"""
+    def __init__(self, name, donations=None):
+        if not donations:
+            donations = []
+
+        self.name = str(name)
+        self._donations = list(donations)
+        self._total = sum(self._donations)
+        self._ave = self._total / len(self._donations) if donations else 0
+
+    @property
+    def donations(self):
+        """Get list of donations for this donor"""
+        return self._donations
+
+    @property
+    def num_donations(self):
+        """Get number of donations given by this donor"""
+        return len(self._donations)
+
+    @property
+    def total_donations(self):
+        """Get total amount of donations given by this donor"""
+        return self._total
+
+    @property
+    def average_donations(self):
+        """Get average donation amount for this donor"""
+        return self._ave
+
+    def add_donation(self, amount):
+        """Add donation to this donor's donation history
+
+        Args:
+            donation (int): donation amount
+        """
+        self._donations.append(amount)
+        self._total += amount
+        self._ave = self._total / self.num_donations
+
+
+class DonorDatabase(defaultdict):
+    """A database of Donors"""
+    def __init__(self, *donors):
+        if not donors:
+            donors = []
+
+        super().__init__(Donor, {d.name: d for d in donors})
+        self.min_col_width = 12
+        self.def_pad = 5
+        self.col_sep = ' | '
+        self.col_1_name = 'Donor Name'
+        self.col_2_name = 'Total Given'
+        self.col_3_name = 'Num Gifts'
+        self.col_4_name = 'Average Gift'
```

Comment:
this would be better as:
`self.cols = ['Donor Name', 'Total Given'.. ]`
then pass into format: `(*self.cols)`